reana_workflow_engine_serial/tasks.py

```diff
@@ -133,53 +136,74 @@ def run_serial_workflow(workflow_uuid, workflow_workspace,
             print('~~~~~~ Publishing step:{0}, cmd: {1},'
                   ' total steps {2} to MQ'.
                   format(step_number, command, len(workflow_json['steps'])))
+            submitted_jobs = {"total": 1, "job_ids": [job_id]}
```

Comment:
When a new job is submitted, the number of planned jobs should diminish by one as well?

Keeping the logic where `submitted + succeeded + failed + planned = total number of steps`:

- In beginning of the run:  0 + 0 + 0 + 8 = 8
- When the first task runs: 1 + 0  + 0 + 7 =  8
- After the first task finished: 0 + 1 + 0 + 7 = 8
- When the second task runs:  1 + 1 + 0 + 6 = 8
- ...
- When the last task runs: 1 + 7 + 0 + 0 = 8
- When the workflow is done: 0 + 8 + 0 + 0 = 9

BTW I would use a terminology such as `running + finished + failed + pending = total` which may be clearer when speaking of tasks...

Reply:
This is because "submitted" and "succeeded" are grammatically so similar that people may think both are aggregating the numbers throughout the workflow lifetime -- while in reality one of them is expressing the momentaneous state in a given time, while the other one is expressing the aggregated count since the workflow start. Expressing "time snapshot" vs "time interval" differently would help to clarify this.

Hence my suggestion to change the grammar to suggested "running" and "finished" which will better express the momentaneous nature of some columns ("-ing") vs the accumulator nature of the latter column ("-ed").

(One could also think of preferring accumulator types and change "planned/pending" to "total". Here is why. If a dynamic workflow adds some steps, the planned or total number would simply increase in time. But most workflows do not add any steps -- hence changing the planned/pending (momentaneous nature) to total (accumulating nature) will reduce the number of SQL queries needed to update the workflow status table columns the as the workflow runs.)  This would give `(running, finished, failed, total)` tuple.)

Reply:
the reason I chose `submitted` is because that is the state that the engine knows. It does know that is has submitted jobs to the job-controller.. whether they are currently running on not is beyond its control (due to queuing etc). If we would want `running` the job controller would need to support an API that gives more detailed information about submitted workloads (e.g. `pending` and `running`)

So in terms of a simple state machine a job for me is

goes through states

```
planned
|
submitted
|         \_______
|                 \
failed          succeeded
```

jobs drop from the top to the bottom and accumulate in failed and succeeded. To me that seems like a quite natural way. Total progress would be

`(failed + succeed) / (failed + succeed + submitted + planned)`

if submitted is confused we can rename it to `in progress`

Reply:
> ultimately i'd be fine with running even though it's a bit imprecise

Well `running` could be called `processing` to be somewhat more precise... As far as the client workflow status reporting goes, we are not concerned so much where in the REANA component sea the task currently stands, only whether it is still "waiting" somewhere in the pipeline or whether it is already being treated ("processing/running").

Anyway the main concern with this PR is elsewhere, notably that the workflow engines should populate these status values in the *exactly same manner*, either as momentaneous snapshots or as aggregations, depending on the column. And its seems that currently Serial and Yadage do these things differently. Hence streamlining this.

> if submitted is confused

I think `submitted` is confusing because "failed" tasks were also "submitted" once, so to speak... One cannot judge based on the name alone which column is showing momentaneous state and which column is showing accumulated numbers from the workflow start. 

This confusion does not occur so much with `running` vs `failed` for example, due to their different tenses; the first implying momentaneous state, the second accumulating state.

Ditto for `planned`. 

E.g. compare:

- N1 planned, N2 succeeded, N3 failed, N4 submitted
- N1 total, N2 finished, N3 failed, N4 running

Reply:
ok I see the point about using `-ing`.. do we all agree that all states other than `failed` and `finished` should be momentaneous states? e.g. pending, processing.. the total would be computed on the fly by `total = finished + failed + pending + processing` and progress is `finished/total`?

Reply:
Yes, "-ed" accumulating, "-ing" momentaneous.

There is a drawback if we use "pending" (when opposed to "total") in that "pending" number will have to be decremented each time the workflow controller submits a new job.  So for static workflows consisting of N jobs in total we will have about N extra SQL queries to do.  Less optimal, but doable. 

An alternative could be to keep `total` in the database. This will be N from the start and the static workflows won't have to touch it so that total=998 always. In case of dynamic workflows the  `total` would be incremented with newly found dynamic count, e.g. originally total=998 foreseen but during execution total=1012 and at the end of execution total=1234.

I am OK with both, provided we consistently treat both engines :)

P.S. In the future we may need to amend this anyway in order to keep the number of SQL queries down. Thinking of ttH workflows consisting of ~500,000 tasks.

Reply:
Does it reduce the number of SQL queries? If we want to track the processing count we need to make an update to the DB everytime we submit a job anyways, no? The same query could be used to decrement pending, right?

Reply:
I agree that for large workflows the update rate is a concern. In yadage we decouple the update intervall for progress tracking from the submissions. So we only track every N seconds.